[WIP] Introspection2

tjhei opened the pull request:
This PR is reworking the Introspection class to be more flexible with the goal to a) simplify the code and b) allow the melt transport (or others) to plug into the creation of the introspection.

The class ``Variable`` and ``IntrospectionBase`` is quite convenient and something that might (at some point?) move into deal.II.

This PR also moves the creation of signals to the beginning of the constructor of the simulator, otherwise plugins can not attach to signals before the FiniteElement system is created.

tjhei commented:
Questions for the reader: does this approach make sense? Any ideas to improve ``Variable`` or ``IntrospectionBase``? Worth putting into deal.II (I know I am using something similar in a different project already)?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+   * sorted collection of finite element variables (of type @p Variable).
+   */
+  template <int dim>
+  class IntrospectionBase
```
Let me digest this whole patch some more, but this already now: I tend to think of a "base" class as one that defines an interface that is then further refined by a derived class. But that's not what you do here. Rather, you "compose" the interface of `Simulator` from the members defined there and the members defined here, so I think that the name "*Base" does not describe this class well. Maybe "VariableDescription"? "Variables"? Something else?

tjhei replied:
Yes, I agree that the naming needs work. It represents a collection of FE variables (basically an ``FESystem``), but it in addition information about blocks (and names, etc.).

We could just call this class ``Introspection`` (in a different namespace), because the ``aspect::Introspection`` basically just adds shortcuts for the ``variable(string)`` access.

jdannberg commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+       * Return the variable with name @p name. Throws an exception if this variable
+       * does not exist.
+       */
+      const Variable<dim> &variable(const std::string &name) const
```
Should we call this something like `variable_for_name`, analogous to the `compositional_index_for_name` that we already have in introspection and that does a similar thing for the index of compositional fields?

jdannberg replied:
I see why you wanted to keep this short when looking at these changes on your melt branch; this function is called quite often. I guess I would at least call it `get_variable` if you think `variable_for_name` is too long.

bangerth replied:
I agree that a more descriptive name would be useful.

jdannberg commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      /**
+       * Return the list of all variables.
+       */
+      const std::vector<Variable<dim> > &variables() const
```
And maybe we could give this function a more specific name too, something like `get_list_of_variables`, or `get_all_variables`?

jdannberg commented on source/simulator/core.cc:
```diff
@@ -75,6 +76,17 @@ namespace aspect
+
+    template <int dim>
+    std::vector<dealii::Variable<dim> > get_variables(const Parameters<dim> &parameters,
+                                                      SimulatorSignals<dim> &signals)
```
Am I getting this right that this function creates the list of variables, and not just returns them?
In this case, should we call it something like `create_list_of_variables` or `set_up_variables`?

bangerth replied:
The function could also use a piece of documentation :-)

I think there is relatively little value in the function, though. It has four lines, and it requires two lines to call it. If you change the signal so that it takes a (const) list of variables and returns a (possibly modified) list of variables, then you can just chain the calls to `construct_variables` and `edit_finite_element_variables` and inline it all into the place where you currently call it.

jdannberg commented:
I haven't looked at everything in detail yet, but I really like this!
The structure of setting up everything in introspection is much nicer now, and from your test it also looks like it is very easy to add new variables. I also think it is fine to use `introspection().variable("something")` instead of the old version, it is still easily readable in the code.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+               const unsigned int multiplicity,
+               const unsigned int n_blocks)
+        : name(name), fe(fe), multiplicity(multiplicity), n_blocks(n_blocks),
+          first_component_index(-1), block_index(-1), base_index(-1)
```
Break this onto one line per variable.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+        Assert(n_blocks == 0
+               || n_blocks == 1
+               || n_blocks == n_components(),
+               ExcMessage("A Variable can only have 0, 1, or n_component number of blocks"));
```
n_component -> n_componentS

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      /**
+       * Constructor for a variable.
+       *
+       * @param name A user-friendly and unique string representation.
```
representation of WHAT?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+       * linear system. A value of 0 will add the next variable to the current
+       * block, while 1 will put the next variable into a new block. A value that is
+       * equal to the number of components will create a block for each
+       * component.
```
This may not matter because few users actually read this. But I don't understand this paragraph because I lack the background of what "blocks" refer to here, how they are used, and how they are used by this class and the classes above it.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+       */
+    private:
+      dealii::FEValuesExtractors::Scalar extractor_scalar_;
+      dealii::FEValuesExtractors::Vector extractor_vector_;
```
I'm not a fan of `variable_`, mostly because we don't use it anywhere else. How about just renaming the variables to `scalar_extractor` and `vector_extractor`?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      std::string name;
+      std_cxx11::shared_ptr<FiniteElement<dim> > fe;
+      unsigned int multiplicity;
+      unsigned int n_blocks;
```
Couldn't you make these variables `const`? Shouldn't they be all private?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      unsigned int n_blocks;
+
+      /**
+       * Initialize this variable as part of a larger FE_System wit the given
```
wit -> with

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+
+      /**
+       * @name Derived quantities
+       * @{
```
What does "derived" here actually mean? And do these variables need to be public and non-const?

bangerth commented:
I think that the `Variable` class is fundamentally a good idea, but I don't like the design at all. The way you currently implement it, the class is initialized in a two-step process: first, the constructor is called and initializes some members; then one needs to call a *second* initialization function that sets up a different set of member variables. Because all member variables are public, it is even possible that there is a *third* phase.

I find this awkward, as well as error prone. My preference would be to come up with a scheme where everything could be initialized in the constructor. I recognize that that involves "counting" what member variables you've already initialized, but oftentimes this can be done by coming up with a helper function that creates objects in a loop and returns a vector of them.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      /**
+         * Construct an empty object.
+         */
+      IntrospectionBase() {}
```
Indentation. Same for the next function.

Maybe make the implementation out-of-line to the class declaration?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+      /**
+       * Fill this object with the given list of @p variables.
+       */
+      void initialize(const std::vector<Variable<dim> > &variables)
```
Does this function have to be public?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+          }
+        Assert(variables_.back().n_blocks != 0
+               || variables_.back().n_components() == 0
+               , ExcMessage("The last variable needs to have >0 blocks."));
```
odd placement of the comma.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+        multiplicities.resize(variables_.size());
+        for (unsigned int i=0; i<variables_.size(); ++i)
+          {
+            fes[i] = &*(variables_[i].fe);
```
variables_[i].fe.get()

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+       */
+      const Variable<dim> &variable(const std::string &name) const
+      {
+        typename std::map<std::string, Variable<dim>*>::const_iterator it
```
make `it` const

bangerth commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,265 @@
+
+      std::map<std::string, Variable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
```
I'm still not a fan of `variable_`. In particular if only *some* variables use this convention.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -170,7 +435,7 @@ namespace aspect
        * that correspond to each of the variables in this problem.
        */
-      const BaseElements base_elements;
+      BaseElements base_elements;
```
That's a shame. I really liked the fact that almost everything in `Introspection` is actually `const`. I think it's a sign of a clean design, and I'd hate to see this go.

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -200,8 +211,8 @@ namespace aspect
-      void call_connector_functions (aspect::SimulatorSignals<2> &signals);
-      void call_connector_functions (aspect::SimulatorSignals<3> &signals);
+      template <int dim>
+      void call_connector_functions (aspect::SimulatorSignals<dim> &signals_);
```
comment is now out of date.

bangerth replied:
(It says "Two functions")

bangerth commented on include/aspect/utilities.h:
```diff
@@ -47,6 +47,27 @@ namespace aspect
+     */
+    inline void split_by_block (std::vector<IndexSet> &partitioned,
+                                const std::vector<types::global_dof_index> &dofs_per_block,
+                                const IndexSet &whole_set)
```
Move the output argument to the end.

The function is expensive. I don't think there's a point in making it `inline`.

bangerth commented on source/particle/world.cc:
```diff
@@ -72,6 +56,8 @@ namespace aspect
                            const unsigned int weight)
     {
+      World<dim>::connect_to_signals(this->get_signals());
+
```
This is, in principle, a change that's completely independent of the rest. Can you submit it separately to separate concerns?

tjhei commented:
> I find this awkward, as well as error prone. My preference would be to come up with a scheme where everything could be initialized in the constructor.

The problem I am trying to tackle with this whole approach is the following:
1. The base code constructs a default set of variables with a) names, b) info needed for constructing an FESystem, c) info about blocks (for linear systems)
2. Plugins/melt code/etc. have a chance to modify this set of variables (remove/replace/change the FE/reorder/etc.)
3. Finally construct the introspection module with this information

To make 2. as easy as possible, I want to hand around a ``std::vector<Variable>`` where fields can be modified easily. This is also the motivation for making it structs with non-const public members.
I decided to reuse ``Variable`` because I felt ``introspection.variable("fluid pressure").component_mask`` is a neat way of allowing access to stuff inside introspection without introducing melt variables into ``Introspection``. I agree that is is less than ideal to have some members inside ``Variable`` be only useful after they exist inside ``Introspection``. The only way I can see to make this work is to not expose the type ``Variable`` inside introspection but to throw them away after initialization. Maybe the syntax would then be ``introspection.component_indices.variable("fluid pressure")``? I find that less than ideal.

tjhei commented:
That said, I have a couple of ideas to improve the interface. Will update this PR.

bangerth commented:
What if you split this into two phases : in the first phase, during which the signal is run, you gather information about what elements the different parts of the program want to have. In the second phase, you build the Variable objects from this intermediate information. This way, you know up front what these objects are going to look like and can build them in their final form. 

As a side note, our code SHOULD be independent of the order in which the base elements go into the system.  (If we do hard code a block number somewhere, then I consider it a bug and we should replace that code by something that properly queries the introspection object for block numbers.) Consequently, I don't really see why a signal should EDIT the list of elements. It should be enough to ADD to it, which could be fine without even giving signals access to the current list.

tjhei commented:
> As a side note, our code SHOULD be independent of the order in which the base elements go into the system.

Note that the order of the variables matters if you want several in a single block (for example for the direct solver). For that the variables need to be behind each other.

I tried it out and we do depend on the order in a few other places right now, which is a bug of course. I was going to tackle that in a separate PR. This is now easy to test, because you can add a dummy variable between each of the current variables. :-)

> Consequently, I don't really see why a signal should EDIT the list of elements. It should be enough to ADD to it, which could be fine without even giving signals access to the current list.

I see several use cases where I want to replace or edit a variable. The DG discretization could have been done as a plugin for example but only if you can modify existing variables. For the melting we need to move variables into different blocks (because we are not solving for "pressure" but compute it as a post-process). Etc.

tjhei commented:
There is still a lot of cleanup left to do (move new classes into separate files, do not do implementation inline, etc.), but I made a bunch of changes based on your feedback. It would be great if you could take another look.

jdannberg commented on include/aspect/introspection.h:
```diff
@@ -29,10 +29,289 @@
+    std::string name;
+    std_cxx11::shared_ptr<FiniteElement<dim> > fe;
+    unsigned int multiplicity;
+    unsigned int n_blocks;
```
Wasn't Wolfgang's idea to make these variables private and const? You do not want to edit them afterwards, correct? All you might want to do is remove a VariableDefinition from the vector that is used to construct introspection, or move it to a different position, but you do not want to change its name or something like that, right?

tjhei replied:
My vote would be to keep it as is. If I can remove an entry and add an entry I can also indirectly modify an entry, right? So I don't mind that we can edit things directly. Maybe you want to change the pressure degree for example...

jdannberg commented on source/simulator/introspection.cc:
```diff
@@ -32,166 +32,157 @@ namespace aspect
       for (unsigned int i=0; i<n_compositional_fields; ++i)
-        ci.compositional_fields.push_back(comp++);
-      return std::make_pair(comp, ci);
+        ci.compositional_fields.push_back(ib.variable("compositions").first_component_index+i);
```
What happens if someone uses the signal to remove one of these default variables, e.g. the pressure? Then all of these `setup_...()` function will trigger an assertion. Is this what we want (we do not allow these variables to be removed) or should we handle this somehow?

tjhei replied:
You are right, right now I assume that these variables exist and an assertion will trigger if they are not there. What do you propose to do if they don't?

jdannberg replied:
@gassmoeller suggested we could put it into a try... catch... block and set it to a signalling NAN if it's not there. Then it would be possible to remove one of the variables if it is not needed, and it would work as long as this variable is not accessed anywhere. I just thought we may want to think about if this is something we want to allow, it's nothing we necessarily have to change.

tjhei replied:
I played with this a bit, but
a) it makes the code somewhat ugly/long 
b) there is no invalid object/NaN for most of these things (invalid ``ComponentMask``?) 
c) things like ``introspection.component_indices.temperature`` are used in so many places in the code, how can we expect ASPECT to run without this? That means a plugin would need to replace every single place...

bangerth replied:
I think this discussion's "I don't know how to do this, there doesn't seem to be a way to express what I want" style mirrors my general unease about the approach. I believe that if you can't quite express what you want, then you've chosen the wrong original design, and now we're trying to shoehorn something into the design that doesn't fit and are not getting anywhere. I know it's awkward to throw away something one has spent a good long time implementing, but this may be one of those cases :-(

I would again like to suggest doing things in a two-step approach where everyone says what they need in some intermediate format, and the second step sets it all up in one central place, in its final form.

tjhei replied:
> I would again like to suggest doing things in a two-step approach where everyone says what they need in some intermediate format, and the second step sets it all up in one central place, in its final form.

This is what I am doing now. Did you see that I have a std::vector<VariableDefinition> as intermediate form and FEVariableSystem (container of FEVariable) as final form?

> I think this discussion's "I don't know how to do this, there doesn't seem to be a way to express what I want" style mirrors my general unease about the approach.

We have three choices: 
1. keep ``introspection.component_mask.velocity`` and always require that there is a variable "velocity" (this is the current state)
2. keep ``introspection.component_mask.velocity`` but initialize to something invalid (that is what Juliane suggested)
3. throw the old introspection away and always write it as ``introspection.variables("velocity").component_mask``. Is that what you want?

I don't like 3 because it removes compile time checks to runtime. I don't think we gain much by 2 over 1 but I am happy to do it if you guys prefer it. Or we do it when needed.

bangerth replied:
On 03/03/2016 05:08 PM, Timo Heister wrote:
>
> 1. keep |introspection.component_mask.velocity| and always require that there
> is a variable "velocity" (this is the current state)
> 2. keep |introspection.component_mask.velocity| but initialize to something
> invalid (that is what Juliane suggested)
> 3. throw the old introspection away and always write it as
> |introspection.variables("velocity").component_mask|. Is that what you want?
>
> I don't like 3 because it removes compile time checks to runtime.

3 requires a run-time lookup in a map or list for every access. We do these in 
the innermost loop, so this seems unappealingly slow.


> I don't
> think we gain much by 2 over 1 but I am happy to do it if you guys prefer it.

Why would we want to use ASPECT for a problem where we don't have a velocity? 
If we assume that we always have one, then 1 seems good enough.

My apologies for being slow to catch up with the current state of the patch -- 
haven't gotten much ASPECT reviewing done lately :-(

jdannberg replied:
I could imagine models where we do not need the temperature, in case it's only a mechanical model. Of course, as Timo already mentioned, right now there are a lot of places in the code where we use things like `introspection.component_indices.temperature`, and the plugin would have to replace them. So I am totally fine with suggestion 1, it was just an idea that I had.

bangerth commented:
On 03/01/2016 05:43 PM, Timo Heister wrote:
>     As a side note, our code SHOULD be independent of the order in which the
>     base elements go into the system.
>
> Note that the order of the variables matters if you want several in a single
> block (for example for the direct solver). For that the variables need to be
> behind each other.

Ah, good point. That only applies to u,p, though, right?


> I tried it out and we do depend on the order in a few other places right now,
> which is a bug of course. I was going to tackle that in a separate PR. This is
> now easy to test, because you can add a dummy variable between each of the
> current variables. :-)

Yes, nice. Thanks for looking into these bugs!


>     Consequently, I don't really see why a signal should EDIT the list of
>     elements. It should be enough to ADD to it, which could be fine without
>     even giving signals access to the current list.
>
> I see several use cases where I want to replace or edit a variable. The DG
> discretization could have been done as a plugin for example but only if you
> can modify existing variables. For the melting we need to move variables into
> different blocks (because we are not solving for "pressure" but compute it as
> a post-process). Etc.

Hm, that seems unfortunate. But ok, if you think that's *necessary* then I'd 
be ok with that. But the editing doesn't have to happen on the *final* form, 
it could also be in an intermediate, more descriptive form of what variables 
everyone wants/needs.

I think the choice between FE_Q and FE_DGQ is properly made at the time when 
they are introduced, not in a later modifying step.

tjhei commented:
>  But the editing doesn't have to happen on the *final* form,
it could also be in an intermediate, more descriptive form of what variables
everyone wants/needs.

This is what I am doing now. Did you see that I have a ``std::vector<VariableDefinition>`` as intermediate form and ``FEVariableSystem`` (container of ``FEVariable``) as final form?

> I think the choice between FE_Q and FE_DGQ is properly made at the time when
they are introduced, not in a later modifying step.

This how it is implemented. I was just giving examples for cases where one might want to edit certain variables. I have another example: I am planning to do a numerical exercise (also melt related), where I don't need the inf-sup condition so I can change the pressure degree.

tjhei commented:
>> Note that the order of the variables matters if you want several in a single
>> block (for example for the direct solver). For that the variables need to be
>> behind each other.
> 
> Ah, good point. That only applies to u,p, though, right?

Yes, except with melt transport. Then it is a little different (u, p_f, and p_c are in the same block, while we shuffle p further back and compute it as a post-process). These are the four cases:
```
[u p] [T] [c1] [c2]  # direct solver
[u] [p] [T] [c1] [c2] # iterative solver
[u p_f p_c] [u_f] [p] [T] [c1] [c2] # melt, direct solver
[u] [p_f p_c] [u_f] [p] [T] [c1] [c2] # melt, iterative solver 
```

tjhei commented:
rebased, squashed, cleaned up more.

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+   * the class FEVariableSystem. It groups the FiniteElement<dim> with other
+   * information like a name and the block structure used in linear systems. A
+   * vector of these structs will be converted into an ordered collection of
+   * objects of type an FEVariable represented by FEVariableSystem.
```
grammar

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+     * component.
+     */
+    VariableDefinition(const std::string &name,
+                       std_cxx11::shared_ptr<FiniteElement<dim> > fe,
```
might as well make it a const reference -- makes no difference in this case, but is possible.

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+     *
+     * @param n_blocks Number of blocks this variable represents inside the
+     * linear system. A value of 0 will add the next variable to the current
+     * block, while 1 will put the next variable into a new block. A value
```
what is the "next" variable here?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+     * linear system. A value of 0 will add the next variable to the current
+     * block, while 1 will put the next variable into a new block. A value
+     * that is equal to the number of components will create a block for each
+     * component.
```
I keep getting confused by components and blocks because I think about situations where someone adds an RT element or similar. Can you maybe say somewhere that the only elements allowed for the second argument are *scalar* (and maybe add an assertion)?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+    /**
+     * Name of the variable used in FEVariableSystem to identify it.
+     */
+    std::string name;
```
make it const. same for all of the other members since you're now setting them in the constructor.

tjhei replied:
If I make them const, you can no longer easily modify the members of this struct when the signal is fired. You can still create a new variable to replace an existing one, but I feel like this is somewhat weird.
If you insist I can do the change (we don't need this for the melting).

bangerth replied:
I like classes with all const members. I think you can achieve this here since I don't see a point in editing an existing variable -- when you edit, what you really say is "I don't need this existing variable, let me replace this with something I *do* need".

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+{
+  /**
+   * Structure to describe everything necessary to define a single variable of
+   * the class FEVariableSystem. It groups the FiniteElement<dim> with other
```
I think what would be useful to say here at the top is that this class defines *a variable in isolation*, whereas FEVariableSystem defines a variable *as part of a collection of variables in an FESystem*. Can you add some such sentence to the documentation?

Realizing this, it also becomes clear to me why I am so confused about the n_blocks variable, since that is in reality something that is not specific to a variable in isolation, but in connection to other variables. Would a better strategy be to not let this member variable be part of the current class, but pass it as a second, independent, argument when combining VariableDefinition objects into a system?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+
+      unsigned int first_component_index;
+      unsigned int block_index;
+      unsigned int base_index;
```
Could all of these now be made `const`?

tjhei replied:
Yes, but then I don't have the default ``operator=()``, which is need to use ``reserve()`` in the ``std::vector``.  Or am I missing something?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+
+    private:
+      dealii::FEValuesExtractors::Scalar scalar_extractor;
+      dealii::FEValuesExtractors::Vector vector_extractor;
```
documentation missing

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+   * information that can be queried.
+   */
+  template<int dim>
+  struct FEVariable: public VariableDefinition<dim>
```
How about FESystemComponent and the next class FESystemComponentCollection?

tjhei replied:
The name "component" is confusing, because one variable can have more than one component (in the FE language). I like the name collection though. "FEVariable" and "FEVariableCollection"?

bangerth replied:
Hm, you're right. "Element" (as in "element of a collection") is also bad.

I didn't like the name `FEVariable` because the name does not make clear any kind of distinction between what a `FEVariable` and a `VariableDefinition` is. They sound the same to me. The difference, semantically, is that the latter stands on its own, whereas the former is an element in a collection. So maybe `FECollectionComponent` or `FECollectionElement`, and then name the next class `FECollection`? Ir `VariableCollectionComponent` and `VariableCollection` (dropping the `FE` altogether)?

tjhei replied:
It now reads VariableDeclaration, FEVariable, FEVariableCollection. I don't think the name for FEVariable is very important because the user doesn't have to deal with them directly.

We could put the two structs (or only FEVariable) within FEVariableCollection, would that be better?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+   * blocks for pressure and each velocity component.
+   */
+  template <int dim>
+  struct VariableDefinition
```
(Not sure about this: Would VariableDescription be a better name? You're describing something, not necessarily defining.)

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+
+  /**
+   * A class that represents a collection of variables (of type FEVariable) to
+   * form a Finite Element system.  A dealii::FE_System can be constructed
```
For historical reasons, the class is actually `FESystem` (no underscore) :-(

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+   * A class that represents a collection of variables (of type FEVariable) to
+   * form a Finite Element system.  A dealii::FE_System can be constructed
+   * from it using get_fes() and get_multiplicities().  Other information
+   * contained in this class on top of FE_System are names, and the block
```
same

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+    protected:
+      std::vector<FEVariable<dim> > variables;
+
+      std::map<std::string, FEVariable<dim>* > name_map;
```
That seems duplicative to store both variables. What if you stored a `std::vector<std::pair<name,FEVariable>>` ? Presumably a linear search through the vector isn't going to be more expensive than a logarithmic one through a `std::map` with 6 elements :-)

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+
+      std::map<std::string, FEVariable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
```
Still not a fan of the underscores. Maybe `number_of_components`, etc.?

tjhei replied:
I prefer ``some_name_`` over
```
int some_name()
{
  return some_totally_different_name;
}
```
because I have to do mental work to associate the variable and the access function with the latter choice.

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+      std::map<std::string, FEVariable<dim>* > name_map;
+      unsigned int n_components_;
+      unsigned int n_blocks_;
+      std::vector<const FiniteElement<dim> *> fes;
```
Or maybe a `std::vector<std::tuple<string,FEVariable,FiniteElement*>>`?

tjhei replied:
This is the variable used to construct the ``FESystem``, so it needs to have this type.

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+      unsigned int n_blocks_;
+      std::vector<const FiniteElement<dim> *> fes;
+      std::vector<unsigned int> multiplicities;
+      std::vector<unsigned int> components_to_blocks;
```
oh, the pair that grew to a tuple grew by two more members.

not sure the tuple is worth it any more at this time. Instead of structure of arrays, one could envision an array of structures, but I'll leave this up to you.

tjhei replied:
Again, ``multiplicities`` is used in the FESystem constructor. ``components_to_blocks`` is used in various ``DofTools`` functions.

bangerth replied:
OK, just make sure you have assertions to ensure these arrays are all the same length.

Maybe add a comment in front of these variables that explain what they stand for?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+                              const unsigned int component_index,
+                              const unsigned int block_index,
+                              const unsigned int base_index)
+    : VariableDefinition<dim> (fe_variable.name,
```
why can't you use the copy constructor of `VariableDefinition` here?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+    base_index (base_index)
+  {
+    if (this->n_components()==1)
+      scalar_extractor.component = component_index;
```
This seems fragile, since you assume that that's the name of the member variable -- which it is, but we don't ever seem to use it outside of `FEValues`. How about
```
  scalar_extractor = FEValuesExtractors::Scalar(component_index);
```
and below use a default constructed object for the else branch?

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+      scalar_extractor.component = -1;
+
+    if (this->n_components()==dim)
+      vector_extractor.first_vector_component = component_index;
```
same here

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,423 @@
+  const dealii::FEValuesExtractors::Scalar &
+  FEVariable<dim>::extractor_scalar() const
+  {
+    Assert(this->n_components()==1, ExcMessage("not a scalar FE"));
```
-> "You cannot ask for the scalar extractor of a non-scalar variable."

bangerth commented on include/aspect/introspection.h:
```diff
@@ -51,23 +56,33 @@ namespace aspect
    */
   template <int dim>
-  struct Introspection
+  struct Introspection: public dealii::FEVariableSystem<dim>
```
Ah, do you really want to put this class into namespace `dealii`? That would seem awkward and a violation of the encapsulation principle. If we ever move it into deal.II itself, we can always clone it and just deprecate the old version.

I'm slightly sad that the nice and self-contained `Introspection` class now consists of parts that one has to look up on different pages. But I guess that's what it is. Unless you want to integrate the FEVariableSystem (which at this point doesn't contain very many things any more) back into `Introspection`.

tjhei replied:
I think of ``FEVariableSystem`` to be independent of ASPECT (and I will use this in other projects and in fact was using an older variant of this already), while ``Introspection`` is certainly ASPECT specific. That is a good enough reason for me to keep them separate.

I can move it out of the dealii namespace for now though.

bangerth replied:
Yes, please. You constrain what we can and cannot do in deal.II if you step on another project's namespace. I'm very much in favor of putting this into the ASPECT namespace, simply because that's the code where it currently lives. I'm not opposed to eventually putting this into deal.II, just like we did with several functions from namespace Utilities, but in that case they also change namespace (as they need to if we want to stay backward compatible).

bangerth commented on include/aspect/introspection.h:
```diff
@@ -51,23 +56,33 @@ namespace aspect
+       * based on the given set of @p parameters.
+       */
+      static std::vector<dealii::VariableDefinition<dim> >
+      construct_variables (const Parameters<dim> &parameters);
```
could this be private? Or just a function in some anonymous namespace?

bangerth commented on include/aspect/introspection.h:
```diff
@@ -190,13 +207,7 @@ namespace aspect
-       * block it corresponds to.
-       */
-      const std::vector<unsigned int> components_to_blocks;
+      const ComponentMasks component_masks;
```
I like that this can now be `const` :-)

tjhei replied:
did you also see how much ugly code got deleted? :-)

bangerth commented on include/aspect/simulator_signals.h:
```diff
@@ -71,6 +71,17 @@ namespace aspect
+     * describes the simulator to act on, and an (output)
+     * argument that represents the finite element variables.
+     */
+    boost::signals2::signal<void (std::vector<dealii::VariableDefinition<dim> > &)>  edit_finite_element_variables;
```
same here with the namespace issue.

bangerth replied:
documentation is no longer current.

bangerth commented on include/aspect/utilities.h:
```diff
@@ -47,6 +47,15 @@ namespace aspect
+     * given by the @p dofs_per_block structure.
+     */
+    void split_by_block (const std::vector<types::global_dof_index> &dofs_per_block,
+                         const IndexSet &whole_set,
```
say something like "The numbers of dofs per block need to add up to the size of the index space described by the second argument."

bangerth commented on source/simulator/introspection.cc:
```diff
@@ -32,166 +32,157 @@ namespace aspect
+      base_elements.velocities = fevs.variable("velocity").base_index;
+      base_elements.pressure = fevs.variable("pressure").base_index;
+      base_elements.temperature = fevs.variable("temperature").base_index;
+      base_elements.velocities = fevs.variable("velocity").base_index;
```
duplicate line to line 80.

tjhei replied:
good catch!

bangerth commented on source/simulator/introspection.cc:
```diff
@@ -32,166 +32,157 @@ namespace aspect
+                                      std_cxx11::shared_ptr<FiniteElement<dim> >(
+                                        new FE_Q<dim>(parameters.stokes_velocity_degree)),
+                                      dim,
+                                      use_direct_solver?0:1)
```
ugh, this is ugly :-( but I don't really see another way, other than what I discussed above about taking this one field out of the structure.

tjhei replied:
if we leave the members non-const we could write this as:
```
variables.push_back(dealii::VariableDefinition<dim>("velocity"));
variables.back().fe = new ...;
variables.back().multiplicity = dim;
variables.back().n_blocks = use_direct_solver?0:1;
```
Do you like that better?

bangerth replied:
We miscommunicated. I didn't object to the creation of a variable inside the call to `push_back` -- that's elegant, and should stay that way. My objection was to the piece `use_direct_solver?0:1` because that is coming out of blue air and at first seems to have nothing to do at all with the creation of a finite element. Might be better to write it as
```
  bool some_descriptive_name = use_direct_solver?0:1;
  variables.push_back (....., some_descriptive_name));
```

tjhei commented:
I think I addressed most comments.

I don't know how to make the members of ``VariableDeclaration`` const because ``std::vector<T>`` requires ``T`` to have ``operator=``. Lucky me? :-)

bangerth commented on include/aspect/fe_variable_system.h:
```diff
@@ -0,0 +1,475 @@
+
+
+  template <int dim>
+  VariableDeclaration<dim>::VariableDeclaration(const std::string &name,
```
None of this is time critical -- put the implementation into a .cc file.

bangerth commented on source/simulator/introspection.cc:
```diff
@@ -32,166 +32,155 @@ namespace aspect
+    const bool use_direct_solver = parameters.use_direct_stokes_solver;
+
+
+    const unsigned int n_velocity_blocks = use_direct_solver?0:1;
```
it took me a long time to see that `use_direct_solver` is used in this line because I though it was still used further down. Just put lines 119 and 122 next to each other to make the connection clearer.

bangerth commented:
OK, after the minor issues mentioned above.

I apologize for taking so long to get through this in review, and for having so many comments. You know that I think that the *design* of software is the most essential part, more so than how it is implemented, and this patch goes at the core of ASPECT. I think the end result is better than where we started. I will admit that I still don't particularly like it, because it is complicated, but I recognize that the complexity allows for the flexibility you need for your melt work and that may help us later on.

Can you please also open a github issue to remind us that we will eventually need to document this system somewhere and somehow.

tjhei commented:
Yeah, I know these kind of PRs are hard to review (and also hard to write...). I hope I got to all of your comments.

bangerth commented:
No doubt they're also hard to write. Thanks a lot for the work and the patience!

tjhei commented:
Any further things for me to do?

bangerth commented:
I think that's it. Squash (or don't) and merge please.

tjhei commented:
Great! Squashed and ready to go if the tests pass.
